File: autoflow.py
#!/usr/bin/env python3
"""
AutoFlow - Automatic website workflow discovery
"""

import asyncio
import json
from urllib.parse import urljoin, urlparse
from playwright.async_api import async_playwright

class AutoFlow:
    def __init__(self, url):
        self.base_url = url
        self.base_domain = urlparse(url).netloc
        self.visited = set()
        self.pages = []  # Discovered pages
        self.requires_login = False
        self.login_form_found = None
        
    async def discover(self, max_pages=20):
        """Discover all pages starting from main URL"""
        
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context()
            page = await context.new_page()
            
            # Handle dialogs (alerts, prompts)
            page.on("dialog", lambda dialog: dialog.dismiss())
            
            await self._visit_page(page, self.base_url)
            
            await browser.close()
            
        return {
            "base_url": self.base_url,
            "pages": self.pages,
            "requires_login": self.requires_login,
            "login_form_found": self.login_form_found
        }
    
    async def _visit_page(self, page, url, parent_menu=None):
        """Visit a page and discover its elements"""
        
        if url in self.visited or len(self.visited) >= 20:
            return
            
        self.visited.add(url)
        print(f"  📄 Visiting: {url}")
        
        try:
            # Navigate to page
            response = await page.goto(url, wait_until="domcontentloaded", timeout=15000)
            
            if response and response.status >= 400:
                print(f"     ❌ Error: {response.status}")
                return
                
        except Exception as e:
            print(f"     ❌ Failed: {str(e)[:50]}")
            return
        
        # Get page title
        title = await page.title() or "Untitled"
        
        # Extract elements
        elements = await self._extract_elements(page)
        
        # Check for login form
        login_detected = await self._check_login_form(page)
        if login_detected:
            self.requires_login = True
            self.login_form_found = url
            print(f"     🔐 Login form detected!")
        
        # Find navigation links
        links = await self._find_menu_links(page)
        
        # Store page info
        page_info = {
            "url": url,
            "name": self._guess_name(url, title),
            "title": title,
            "elements": elements,
            "links_found": len(links),
            "is_login_page": login_detected
        }
        self.pages.append(page_info)
        
        # Visit linked pages (limited)
        for link in links[:5]:  # Max 5 links per page
            if link not in self.visited:
                await self._visit_page(page, link)
    
    async def _extract_elements(self, page):
        """Extract meaningful elements from page"""
        
        elements = []
        
        # Common element types to look for
        selectors = {
            "buttons": ["button", "a.button", "input[type=submit]", ".btn"],
            "inputs": ["input", "textarea", "select"],
            "links": ["a"],
            "headings": ["h1", "h2", "h3"],
            "cards": [".card", ".product", ".item", "article"],
            "forms": ["form"],
            "tables": ["table"],
            "images": ["img"]
        }
        
        for category, selector_list in selectors.items():
            for selector in selector_list:
                try:
                    els = await page.query_selector_all(selector)
                    if els:
                        count = len(els)
                        if count > 0:
                            elements.append({
                                "category": category,
                                "selector": selector,
                                "count": count,
                                "examples": await self._get_element_texts(els[:3])
                            })
                        break  # Only first matching selector
                except:
                    pass
        
        return elements
    
    async def _get_element_texts(self, elements):
        """Get text content from elements"""
        texts = []
        for el in elements:
            try:
                text = await el.text_content()
                if text and len(text.strip()) > 0:
                    texts.append(text.strip()[:50])
            except:
                pass
        return texts
    
    async def _check_login_form(self, page):
        """Check if page has login form"""
        
        login_indicators = [
            "password", "login", "signin", "email", 
            "username", "entrar", "acessar", "logar"
        ]
        
        # Check for password input
        password_input = await page.query_selector("input[type=password]")
        if password_input:
            return True
        
        # Check form labels
        html = await page.content().lower()
        for indicator in login_indicators:
            if indicator in html:
                # Check if near a form
                form = await page.query_selector("form")
                if form:
                    return True
        
        return False
    
    async def _find_menu_links(self, page):
        """Find navigation links on page"""
        
        links = []
        
        # Look for navigation elements
        nav_selectors = ["nav", "header", ".menu", ".nav", ".navbar", "[role=navigation]"]
        
        for selector in nav_selectors:
            try:
                nav = await page.query_selector(selector)
                if nav:
                    anchors = await nav.query_selector_all("a")
                    for a in anchors:
                        href = await a.get_attribute("href")
                        if href:
                            full_url = urljoin(self.base_url, href)
                            if urlparse(full_url).netloc == self.base_domain:
                                if full_url not in links:
                                    links.append(full_url)
            except:
                pass
        
        # If no nav found, look for links in main content
        if not links:
            anchors = await page.query_selector_all("main a, .content a, body a")
            for a in anchors[:10]:
                href = await a.get_attribute("href")
                if href and not href.startswith("#"):
                    full_url = urljoin(self.base_url, href)
                    if urlparse(full_url).netloc == self.base_domain:
                        if full_url not in links:
                            links.append(full_url)
        
        return links
    
    def _guess_name(self, url, title):
        """Guess a friendly name for the page"""
        
        path = urlparse(url).path
        if path == "/" or path == "":
            return "Home"
        
        # Get last part of path
        parts = [p for p in path.split("/") if p]
        if parts:
            name = parts[-1]
            # Clean up
            name = name.replace("-", " ").replace("_", " ")
            name = name.title()
            return name
        
        return title[:30] if title else "Page"


async def main():
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python autoflow.py <url>")
        print("Example: python autoflow.py https://example.com")
        sys.exit(1)
    
    url = sys.argv[1]
    print(f"🔍 AutoFlow - Discovering: {url}")
    print("=" * 50)
    
    autoflow = AutoFlow(url)
    result = await autoflow.discover()
    
    print("=" * 50)
    print(f"✅ Discovery complete!")
    print(f"   Pages found: {len(result['pages'])}")
    print(f"   Login required: {result['requires_login']}")
    
    # Save to JSON
    output_file = "workflow.json"
    with open(output_file, "w") as f:
        json.dump(result, f, indent=2)
    print(f"   Saved to: {output_file}")
    
    # Print summary
    print("\n📋 Pages discovered:")
    for i, page in enumerate(result["pages"], 1):
        print(f"   {i}. {page['name']} ({page['url']})")
        print(f"      Elements: {', '.join([e['category'] for e in page['elements']])}")


if __name__ == "__main__":
    asyncio.run(main())

File: autoflow.js
const { chromium } = require('playwright');

class AutoFlow {
    constructor(url) {
        this.baseUrl = url;
        this.visited = new Set();
        this.pages = [];
    }

    async discover(maxPages = 30) {
        this.maxPages = maxPages;
        const browser = await chromium.launch({ headless: true });
        const context = await browser.newContext();
        const page = await context.newPage();

        // Start with the base URL
        await this.visitPage(page, this.baseUrl);
        
        // Try to find and click menu items
        await this.scanMenus(page);

        await browser.close();
        
        return {
            baseUrl: this.baseUrl,
            pages: this.pages,
            menuScanned: true
        };
    }

    async visitPage(page, url) {
        if (this.visited.has(url) || this.pages.length >= this.maxPages) return;
        
        this.visited.add(url);
        console.log(`   📄 Visiting: ${url}`);
        
        try {
            await page.goto(url, { waitUntil: 'networkidle', timeout: 15000 });
            
            // Wait a bit for any dynamic content
            await page.waitForTimeout(1000);
            
            // Extract all relevant elements with their selectors
            const elements = await this.extractElements(page);
            
            // Find menu links on this page
            const menuLinks = await this.findMenuLinks(page);
            
            const pageData = {
                url: url,
                name: this.guessName(url, await page.title()),
                title: await page.title(),
                elements: elements,
                menuLinks: menuLinks,
                selectors: this.buildSelectorMap(elements)
            };
            
            this.pages.push(pageData);
            
            // Visit pages found in menus
            for (const link of menuLinks.slice(0, 10)) { // Limit to avoid too many pages
                if (this.pages.length >= this.maxPages) break;
                await this.visitPage(page, link);
            }
            
        } catch (e) {
            console.log(`      ⚠️ Error: ${e.message.split('\n')[0]}`);
        }
    }

    async extractElements(page) {
        const elementCategories = [
            { category: 'buttons', selector: 'button, [role="button"], input[type="submit"], input[type="button"]' },
            { category: 'inputs', selector: 'input[type="text"], input[type="email"], input[type="search"], input[type="password"], textarea' },
            { category: 'links', selector: 'a[href]' },
            { category: 'selects', selector: 'select' },
            { category: 'forms', selector: 'form' },
            { category: 'headings', selector: 'h1, h2, h3, h4, h5, h6' },
            { category: 'images', selector: 'img' },
            { category: 'lists', selector: 'ul, ol, [role="listbox"]' },
            { category: 'tables', selector: 'table, [role="table"]' },
            { category: 'dialogs', selector: '[role="dialog"], modal, .modal' },
            { category: 'menus', selector: 'nav, [role="navigation"], .menu, .nav, header' }
        ];

        const elements = [];
        
        for (const { category, selector } of elementCategories) {
            try {
                const els = await page.$$(selector);
                for (const el of els) {
                    const selector = await this.getUniqueSelector(page, el);
                    const text = await el.textContent().catch(() => '');
                    const tag = await el.evaluate(e => e.tagName);
                    
                    if (text && text.trim()) {
                        elements.push({
                            category,
                            tag: tag.toLowerCase(),
                            selector: selector,
                            text: text.trim().substring(0, 100),
                            visible: await el.isVisible().catch(() => true)
                        });
                    }
                }
            } catch (e) {
                // Ignore selector errors
            }
        }
        
        return elements;
    }

    async getUniqueSelector(page, element) {
        // Try to get a human-readable selector
        const id = await element.getAttribute('id');
        if (id) return `#${id}`;
        
        const classes = await element.getAttribute('class');
        if (classes) {
            const classList = classes.split(/\s+/).filter(c => c);
            if (classList.length > 0) {
                const tag = await element.evaluate(e => e.tagName);
                return `${tag.toLowerCase()}.${classList[0]}`;
            }
        }
        
        // Fall back to CSS selector
        try {
            return await element.evaluate((el) => {
                if (el.id) return `#${el.id}`;
                if (el.className && typeof el.className === 'string') {
                    return `${el.tagName.toLowerCase()}.${el.className.split(' ')[0]}`;
                }
                return el.tagName.toLowerCase();
            });
        } catch (e) {
            return 'unknown';
        }
    }

    async findMenuLinks(page) {
        const links = new Set();
        
        try {
            // Find all navigation menus
            const menus = await page.$$('nav, [role="navigation"], .menu, .nav, header, .navbar, .header');
            
            for (const menu of menus) {
                const menuLinks = await menu.$$('a[href]');
                for (const link of menuLinks) {
                    const href = await link.getAttribute('href');
                    const text = await link.textContent();
                    
                    if (href && !href.startsWith('#') && !href.startsWith('javascript')) {
                        const fullUrl = href.startsWith('http') ? href : new URL(href, this.baseUrl).href;
                        if (fullUrl.startsWith(this.baseUrl)) {
                            links.add(fullUrl);
                        }
                    }
                }
            }
            
            // Also try dropdown menus
            const dropdowns = await page.$$('[role="menu"], .dropdown-menu, .menu-dropdown, [class*="dropdown"]');
            for (const dropdown of dropdowns) {
                const ddLinks = await dropdown.$$('a[href]');
                for (const link of ddLinks) {
                    const href = await link.getAttribute('href');
                    if (href && !href.startsWith('#')) {
                        const fullUrl = href.startsWith('http') ? href : new URL(href, this.baseUrl).href;
                        if (fullUrl.startsWith(this.baseUrl)) {
                            links.add(fullUrl);
                        }
                    }
                }
            }
            
        } catch (e) {
            // Ignore errors
        }
        
        return Array.from(links);
    }

    async scanMenus(page) {
        console.log(`\n🗺️  Scanning menus...`);
        
        try {
            // Find all menu items with dropdowns
            const menuItems = await page.$$('[role="menuitem"], .dropdown-toggle, .has-dropdown, [class*="menu-item"]:has(a), li:has(.dropdown)');
            
            for (const item of menuItems) {
                try {
                    // Hover to open dropdown
                    await item.hover();
                    await page.waitForTimeout(500);
                    
                    // Find links in the opened dropdown
                    const dropdownLinks = await item.$$('[role="menuitem"] a, .dropdown-menu a, .submenu a, [class*="dropdown"] a');
                    
                    for (const link of dropdownLinks) {
                        const href = await link.getAttribute('href');
                        const text = await link.textContent();
                        
                        if (href && !href.startsWith('#')) {
                            const fullUrl = href.startsWith('http') ? href : new URL(href, this.baseUrl).href;
                            console.log(`      📂 Menu: ${text?.trim() || 'unknown'} -> ${fullUrl}`);
                            await this.visitPage(page, fullUrl);
                        }
                    }
                } catch (e) {
                    // Ignore individual menu errors
                }
            }
        } catch (e) {
            console.log(`      ⚠️ Menu scan: ${e.message.split('\n')[0]}`);
        }
    }

    buildSelectorMap(elements) {
        const selectors = {};
        for (const el of elements) {
            if (!selectors[el.category]) {
                selectors[el.category] = [];
            }
            selectors[el.category].push({
                selector: el.selector,
                text: el.text
            });
        }
        return selectors;
    }

    guessName(url, title) {
        const path = new URL(url).pathname;
        if (!path || path === '/') return 'Home';
        
        const parts = path.split('/').filter(p => p);
        if (parts.length > 0) {
            return toTitleCase(parts[parts.length - 1].replace(/[-_]/g, ' ').replace(/\.\w+$/, ''));
        }
        
        return title.substring(0, 30);
    }
}

function toTitleCase(str) {
    return str.replace(/\b\w/g, l => l.toUpperCase());
}

// Usage
const url = process.argv[2] || 'https://example.com';
console.log(`🔍 AutoFlow - Menu-Aware Discovery: ${url}\n`);

new AutoFlow(url).discover().then(result => {
    console.log(`\n✅ Discovery complete!`);
    console.log(`   Pages found: ${result.pages.length}`);
    
    // Show summary
    console.log(`\n📋 Summary:`);
    for (const page of result.pages.slice(0, 10)) {
        const categories = page.elements.map(e => e.category).filter((v, i, a) => a.indexOf(v) === i);
        console.log(`   - ${page.name}: ${categories.join(', ')}`);
    }
    
    // Save results
    require('fs').writeFileSync('workflow.json', JSON.stringify(result, null, 2));
    console.log(`\n💾 Saved to workflow.json`);
}).catch(e => {
    console.error(`❌ Error: ${e.message}`);
    process.exit(1);
});
